### PR TITLE
feat(authenticator): Upgrade to Gen2 Amplify version

### PR DESCRIPTION
**Pull request by mattcreaser**
- [x] PR conforms to [Pull Request](https://github.com/aws-amplify/amplify-ui-android/blob/main/CONTRIBUTING.md#contributing-via-pull-requests) guidelines.

*Issue #, if available:*

*Description of changes:*
Changes how Authenticator gets its configuration values to directly use the `AuthConfiguration` exposed by Amplify (added in version 2.15.0). This removes the need to parse JSON in Authenticator.

The main difference is that `PasswordCriteria` is called `PasswordProtectionSettings` in Amplify's AuthConfiguration.

*How did you test these changes?*
(Please add a line here how the changes were tested)

*Documentation update required?*
- [x] No
- [ ] Yes (Please include a PR link for the documentation update)

*General Checklist*
- [ ] Added Unit Tests
- [ ] Added Integration Tests
- [ ] Security oriented best practices and standards are followed (e.g. using input sanitization, principle of least privilege, etc)

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

File: samples/authenticator/app/build.gradle [hunk 1]
```diff
@@ -53,6 +54,7 @@ dependencies {
     // Use this to use published version of Amplify UI
     implementation "com.amplifyframework.ui:authenticator:$authenticatorVersion"
 
+    coreLibraryDesugaring 'com.android.tools:desugar_jdk_libs:2.0.3'
```
Comment (tylerjroach) [inline samples/authenticator/app/build.gradle:57]: Wasn't this v2 causing amplify issues? Want to be careful about bumping here and then transitively causing build failures on amplify

File: authenticator/src/main/java/com/amplifyframework/ui/authenticator/states/StepStateFactory.kt [hunk 2]
```diff
@@ -136,7 +137,7 @@ internal class StepStateFactory(
         deliveryDetails: AuthCodeDeliveryDetails?,
         onSubmit: suspend (newPassword: String, confirmationCode: String) -> Unit
     ) = PasswordResetConfirmStateImpl(
-        passwordCriteria = authConfiguration.passwordCriteria,
+        passwordProtectionSettings = authConfiguration.passwordProtectionSettings!!,
```
Comment (tylerjroach) [inline authenticator/src/main/java/com/amplifyframework/ui/authenticator/states/StepStateFactory.kt:140]: any alternative to `!!`?

File: authenticator/src/main/java/com/amplifyframework/ui/authenticator/states/StepStateFactory.kt [hunk 3]
```diff
@@ -107,7 +108,7 @@ internal class StepStateFactory(
     ) = SignUpStateImpl(
         signInMethod = authConfiguration.signInMethod,
         signUpAttributes = authConfiguration.signUpAttributes,
-        passwordCriteria = authConfiguration.passwordCriteria,
+        passwordProtectionSettings = authConfiguration.passwordProtectionSettings!!,
```
[the remark this answers is not in the document]
Comment (mattcreaser) [reply authenticator/src/main/java/com/amplifyframework/ui/authenticator/states/StepStateFactory.kt:111]: We know this is non-null from the check in `AuthProvider.getConfiguration`. The alternative is to keep our own version of `AuthConfiguration` which has this field be non-nullable. Will consider doing that.

File: samples/authenticator/app/build.gradle (hunk 1 above)
Comment (mattcreaser) [reply samples/authenticator/app/build.gradle:57]: Version 2+ requires JDK 11 and Gradle 7.4+. So there's no issue with Amplify it's about the build environment.

**Review by tylerjroach — DISMISSED** [review]

**Review by tylerjroach — APPROVED** [review]

**Review by joon-won — APPROVED** [review]

**Pull request merged**